README.md
# Metsuri, a serial logger

Metsuri is a base component in a serial logging system. It makes it easy to 
monitor hardware that has an output that can be attached with, for example, a 
USB serial interface. 

As the monitoring hardware you will need something that runs the necessary
Python components. For example, a recent RaspberryPi will work splendidly.
16GB SD card is enough, probably a lot smaller will work, as the system tries
to upload the data to the cloud as soon as it is received.

The system comprises three main components, `serial-logger`, 
`log-uploader` and `log-download`. The `serial-logger` attaches to a serial 
line and logs the received data to a file, rotating files with requested 
intervals, `log-uploader` uploads the received events to AWS 
CloudWatch, and you download the logs from the desired internal from the 
cloud with `log-download`.

It is possible to just log the data locally, but AWS CloudWatch makes it
simple to share the updated logs with a team of people. It is also a very 
secure way to store the logs if the loggers are installed in a remote
location.

## Install needed packages

```shell script
% sudo apt-get update
% sudo apt-get install picocom supervisor python3-venv
```

## Create Python env

```shell script
% python3 -m venv ve
% . ve/bin/activate
(ve) % pip install metsuri-*.whl
```

## Copy supervisor config

```shell script
% cd supervisord
% sudo cp *.conf /etc/supervisor/conf.d
```

Currently `log_uploader.conf` needs to be modified. 
Modify the command-line to use correct serial device, log group and log 
stream for the device. Each device should have its own log stream.

If you wish to test the configuration, you can attach the USB serial and call

```shell script
% sudo supervisorctl reload
```

which will start services. Both services should end up in `RUNNING` when
observed with 

```shell script
% sudo supervisorctl status
```

In case of problems, logs can be found under `/var/log/supervisor/`

## Setup connection configuration for AWS

The device should include credentials to allow logging into a CloudWatch log 
stream.

The `~/.aws/config` should include the `[default]` section to point out the
correct region for the CloudWatch target. Alternatively, you can specify, 
for example, `AWS_PROFILE` environment variable among others. 
See [AWS documentation](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html) for details. 

```ini
[default]
output = json
region = eu-central-1
```

By default, the credentials are found in `~/.aws/credentials`, and the 
location can be overridden with `AWS_SHARED_CREDENTIALS_FILE` environment 
variable.

Example `/home/pi/.aws/credentials` below.

```ini
[default]
aws_access_key_id=foo
aws_secret_access_key=bar
```

## Fetching the logs from CloudWatch

In the following example we gather all logs from log stream with names 
starting with `example` from the log group `EXAMPLE-LOG-GROUP` downloading them to files in  
`output-log-dir`.

This method reflects the state of the device logs exactly, but can be heavy
with lots of stream or log entries. Use `--interval` command-line option to
limit the amount of data, or specify a narrow time window with `--from` and 
`--to`. 

```shell script
AWS_SHARED_CREDENTIALS_FILE=creds log-download EXAMPLE-LOG-GROUP example output-log-dir
```


ChangeLog.md
# ChangeLog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.5.0] - 2023-02-14

### Changed

- Project hosted in GitHub as an open source project.
- License is MIT.

### Fixed

- Tolerate timestamps that do not parse with datetime. These can occur if logger loses power during writing of log lines, causing huge year values, for example.

## [0.4.0] - 2022-05-13

### Added

- Requested log stream will be created if it did not exist already.
- Added `--version` to `log-uploader`.

## [0.3.0] - 2022-03-11

### Added

- Added `--ignore-timestamp` to `log-uploader`, which can be used to re-upload a log file even if the corresponding `*.lus` file would make the contents to be otherwise skipped. Mostly for testing.
- Added `log-generate` helper to generate current log files from arbitrary text files.
- Added `log-check` helper to iterate through a log file to check for possible errors.

### Fixed

- Allow starting the `serial-logger` without an attached USB cable. The logger will start logging immediately once the USB is attached.
- Allow starting the log-uploader with a non-existent log file with `--watch`.
- Allow using a log-stream name with common prefix that exactly matches a single file, e.g., `customer-name` and `customer-name-group`.

## [0.2.2] - 2021-10-06

### Fixed

- Fix issue with trying to upload batches larger than AWS PutLogEvents max batch size.
- Be more graceful with ill-formatted log event lines, with, e.g., missing timestamps.
- Fix issue with trying to upload batch with events spanning more than 24 hours.

## [0.2.1] - 2021-04-22

### Fixed
- Attempt to fix an issue causing logger to get stuck while logging without any indication of error. At least should see from the internal log where it gets stuck.

### Changed
- Use `rich` package for nicer progress output for `log-download`.

## [0.2.0]

### Changed
- The whole thing is now a self-contained package, allowing to be installed in a virtual environment, hopefully making setup easier in the future.

### Added
- Added `tox` setup to allow running tests against different python versions.

### Fixed
- Do not choke on non-unicode characters in log file. Catch exceptions in log file reader process and pass exceptions to main process for handling.

## [0.1.1]

### Fixed
- Fixed to work with python 3.7, the default in the current Raspbian
- Fixed USB detach/attach functionality to be lot more robust

## [0.1.0]

### Changed
- If one of `--to` or `--from` is passed alongside `--interval`, the value of `--interval` is used to determine the other end of the time period. Behavior stays the same if both `--to` and `--from` are provided; in that case `--interval` is ignored.

### Added
- Added `--verbose` option to `serial_logger.py` and `log_uploader.py`.
- `log_download.py list-streams` command to show possible log streams to download.
- Added this changelog.

### Fixed
- Open files with 'utf-8' codec on Windows with `log_download.py` to better support Unicode input.
- In case log is written fairly seldom, timeout and send smaller batches even if log file has not been appended.


tests/conftest.py
import pytest
import tempfile
import os


TESTDIR = os.path.dirname(__file__)


@pytest.fixture
def log_file_name():
    with tempfile.TemporaryDirectory(dir=TESTDIR) as temp_dir:
        yield os.path.join(temp_dir, "test.log")


tests/test_logging.py
from freezegun import freeze_time
import os
import datetime
from metsuri.serial_logger import LogFile, collect_serial_debug
from metsuri.log_uploader import get_log_entries, get_timestamp
import unittest.mock as mock
import pytest
import serial


def test_logfile(log_file_name):
    assert not os.path.exists(log_file_name)
    with LogFile(log_file_name):
        pass
    lines = open(log_file_name).readlines()
    assert len(lines) == 2
    assert "started" in lines[0]
    assert "stopped" in lines[1]


@pytest.mark.xfail
def test_freezegun():
    with freeze_time(datetime.datetime(year=2020, month=1, day=1,
                                       microsecond=400000),
                     tz_offset=4) as frozen:
        dt1 = datetime.datetime.now(tz=datetime.timezone.utc)
        dt2 = datetime.datetime.utcnow().replace(tzinfo=datetime.timezone.utc)
        assert dt1 == dt2


def test_timestamps(log_file_name):
    with freeze_time(datetime.datetime(year=2020, month=1, day=1,
                                       microsecond=400000)) as frozen:

        with LogFile(log_file_name) as log:
            frozen.tick()
            log.write_line("foo")
            frozen.tick()
            log.write_line("bar")
    lines = open(log_file_name).readlines()
    assert len(lines) == 4
    assert "foo" in lines[1]
    assert "bar" in lines[2]
    assert lines[1].startswith("2020-01-01T00:00:01.400+00:00")

    entries = list(get_log_entries(log_file_name))
    assert len(entries) == 4
    assert entries[1][0] == datetime.datetime(year=2020, month=1, day=1,
                                              hour=0, second=1,
                                              microsecond=400000, tzinfo=datetime.timezone.utc)


def test_upgraded_logfile(log_file_name):
    log_data = """\
    2020-01-01 00:00:01.400+00:00 foo
    [2020-01-01 00:00:01.500+00:00]: bar
    [2020-01-01T00:00:01.600+00:00]: qux
    2020-01-01T00:00:01.600: zap
    2020-01-01T00:00:01.700 zappa"""
    with open(log_file_name, "w") as fp:
        fp.write(log_data)
        fp.close()

    with freeze_time(datetime.datetime(year=2020, month=1, day=1,
                                       microsecond=400000), tz_offset=5) as frozen:
        entries = list(get_log_entries(log_file_name))

    assert len(entries) == 5
    assert "foo" in entries[0].message

    local_tzinfo = datetime.datetime(year=2020, month=1, day=1).astimezone().tzinfo

    assert entries[3][0] == datetime.datetime(year=2020, month=1, day=1,
                                              hour=0, minute=0, second=1,
                                              microsecond=600000,
                                              tzinfo=local_tzinfo)
    assert entries[4][0] == datetime.datetime(year=2020, month=1, day=1,
                                              hour=0, minute=0, second=1,
                                              microsecond=700000,
                                              tzinfo=local_tzinfo)


def test_upgraded_logfile_and_timestamp(log_file_name):
    log_data = """\
    2020-01-01 00:00:01.400+00:00 foo
    [2020-01-01 00:00:01.500+00:00]: bar
    [2020-01-01T00:00:01.550+00:00]: qux
    2020-01-01T04:00:01.600: zap
    2020-01-01T04:00:01.700 zappa dappa"""
    with open(log_file_name, "w") as fp:
        fp.write(log_data)

    entries = list(get_log_entries(log_file_name))

    assert [e[1] for e in entries] == ["foo", "bar", "qux", "zap",
                                       "zappa dappa"]

    with open(log_file_name + ".lus", "w") as fp:
        fp.write("2020-01-01T03:00:01.599")

    with freeze_time(datetime.datetime(year=2020, month=1, day=1,
                                       microsecond=400000)):
        entries = list(get_log_entries(log_file_name))

    assert len(entries) == 2
    assert [e[1] for e in entries] == ["zap", "zappa dappa"]

    local_tzinfo = datetime.datetime(year=2020, month=1, day=1).astimezone().tzinfo

    assert entries[0][0] == datetime.datetime(year=2020, month=1, day=1,
                                              hour=4, minute=0, second=1,
                                              microsecond=600000,
                                              tzinfo=local_tzinfo)
    assert entries[1][0] == datetime.datetime(year=2020, month=1, day=1,
                                              hour=4, minute=0, second=1,
                                              microsecond=700000,
                                              tzinfo=local_tzinfo)


def test_get_timestamp(log_file_name):
    with open(log_file_name + ".lus", "w") as fp:
        fp.write("2020-01-01T03:00:01.599+00:00")
    ts = get_timestamp(log_file_name)
    assert ts == datetime.datetime(year=2020, month=1, day=1,
                                   hour=3, minute=0, second=1,
                                   microsecond=599000,
                                   tzinfo=datetime.timezone.utc)


def test_write_line_returns_written(log_file_name):
    with freeze_time(datetime.datetime(year=2020, month=1, day=1),
                     tz_offset=3) as frozen:
        with LogFile(log_file_name) as log:
            written = log.write_line("foo")
            assert "2020-01-01T03:00:00" in written
            assert "foo" in written


def test_linefeeds(log_file_name):
    with LogFile(log_file_name) as log:
        log.write_line("foo\n")

    lines = list(map(lambda l: l.strip(), open(log_file_name).readlines()))
    assert all(lines)
    assert len(lines) == 3


def test_flushing(log_file_name):
    with mock.patch('metsuri.serial_logger.open', create=True) as mock_open:
        mock_open.return_value.tell.return_value = 0

        with LogFile(log_file_name) as log:
            # should be called when opening the log.
            mock_open.return_value.flush.assert_called()
            mock_open.return_value.flush.reset_mock()
            mock_open.return_value.flush.assert_not_called()

            log.write_line("foo\n")
            mock_open.return_value.flush.assert_called()


def test_closing(log_file_name):
    with mock.patch('metsuri.serial_logger.open', create=True) as mock_open:
        mock_open.return_value.tell.return_value = 0

        with LogFile(log_file_name) as log:
            log.write_line("foo\n")
            mock_open.return_value.close.assert_not_called()
        mock_open.return_value.close.assert_called_once()


def test_serial_logging(log_file_name):
    with mock.patch('serial.Serial') as MockSerial:
        MockSerial.return_value.readline.side_effect = \
            "foo\n".encode("ascii"), "bar\n".encode("ascii"), \
            "baz\n".encode("ascii"), \
            StopIteration("foo")
        with LogFile(log_file_name) as log:
            collect_serial_debug("/dev/ttyUSB0", log, True)


def test_serial_logging_survives_disconnecting_usb(log_file_name):
    with mock.patch('serial.Serial') as MockSerial, \
            mock.patch('time.sleep') as mock_sleep:
        MockSerial.name = "class"
        initial_serial = mock.MagicMock()
        initial_serial.name = "initial"
        initial_serial.readline = mock.MagicMock(side_effect=[
            "foo\n".encode("ascii"),
            serial.SerialException("foo"),
            StopIteration("foo")
        ])
        last_serial = mock.MagicMock()
        last_serial.name = "last"
        last_serial.readline.side_effect = [
            "bar\n".encode("ascii"),
            StopIteration("foo")
        ]
        MockSerial.side_effect = [initial_serial,
                                  serial.SerialException("fail1"),
                                  serial.SerialException("fail2"),
                                  serial.SerialException("fail3"),
                                  last_serial]
        with LogFile(log_file_name) as log:
            collect_serial_debug("/dev/ttyUSB0", log)
            mock_sleep.assert_called()
        lines = open(log_file_name).readlines()
        # both "foo" and "bar" need to be found
        assert "foo" in lines[2]
        # Log event for disconnection and reconnection emitted
        assert "USB disconnected" in lines[3]
        assert len(list(filter(lambda xx: "USB disconnected" in xx, lines))) == 1
        assert "USB connected" in lines[4]
        assert "bar" in lines[5]


def test_serial_logging_with_stdout(log_file_name):
    with mock.patch('serial.Serial') as MockSerial, \
            mock.patch('metsuri.serial_logger.print') as mock_print:
        MockSerial.return_value.readline.side_effect = \
            "foo\n".encode("ascii"), "bar\n".encode("ascii"), \
            "baz\n".encode("ascii"), StopIteration("foo")
        with LogFile(log_file_name, copy_to_stdout=True) as log:
            collect_serial_debug("/dev/ttyUSB0", log)
        print_arg = mock_print.call_args_list[1][0][0]
        assert isinstance(print_arg, str)
        assert "USB connected" in print_arg
        print_arg = mock_print.call_args_list[2][0][0]
        assert "foo" in print_arg


def test_log_rotation(log_file_name):
    with freeze_time(datetime.datetime(year=2020, month=1, day=1),
                     tz_offset=3) as frozen:
        open(log_file_name + ".foo", "w").write("foo")
        with LogFile(log_file_name, rotation_interval=datetime.timedelta(days=1)) as log:
            log.write_line("foo 1")
            frozen.tick(delta=datetime.timedelta(hours=12))
            log.write_line("foo 2")
            frozen.tick(delta=datetime.timedelta(hours=13))
            log.write_line("foo 3")
            log.write_line("foo 4")
            log.write_line("foo 5")
    assert os.path.exists(log_file_name)
    assert "foo 2" not in open(log_file_name).read()
    # log should have been rotated.
    assert os.path.exists(log_file_name + ".1")
    assert not os.path.exists(log_file_name + ".2")


def test_log_rotation_get_timestamp_on_reopen(log_file_name):
    with freeze_time(datetime.datetime(year=2020, month=1, day=1),
                     tz_offset=3) as frozen:
        open(log_file_name + ".foo", "w").write("foo")
        with LogFile(log_file_name, rotation_interval=datetime.timedelta(days=1)) as log:
            log.write_line("foo 1")
            frozen.tick(delta=datetime.timedelta(hours=12))
            log.write_line("foo 2")
            frozen.tick(delta=datetime.timedelta(hours=11))
            log.write_line("foo 3")

        assert not os.path.exists(log_file_name + ".1")
        with LogFile(log_file_name, rotation_interval=datetime.timedelta(days=1)) as log:
            frozen.tick(delta=datetime.timedelta(minutes=59))
            log.write_line("foo 4")
            assert not os.path.exists(log_file_name + ".1")
            frozen.tick(delta=datetime.timedelta(minutes=2))
            log.write_line("foo 5")

            assert os.path.exists(log_file_name + ".1")


def test_log_rotation_num_retained(log_file_name):
    with freeze_time(datetime.datetime(year=2020, month=1, day=1),
                     tz_offset=3) as frozen:
        with LogFile(log_file_name, rotation_interval=datetime.timedelta(hours=1),
                     num_retained_logfiles=3) as log:
            for ii in range(10):
                frozen.tick(delta=datetime.timedelta(minutes=30))
                log.write_line(f"foo {ii}")
    assert os.path.exists(log_file_name)
    assert "foo 9" in open(log_file_name).read()
    # log should have been rotated.
    assert os.path.exists(log_file_name + ".3")
    assert not os.path.exists(log_file_name + ".4")



src/metsuri/log_check.py
"""
Usage: log-check [options] LOG_FILE

Options:
  --verbose            Enable verbose logging.

"""
import docopt
import logging
from metsuri.log_uploader import get_log_entries

logger = logging.getLogger(__name__)


def main():
    opts = docopt.docopt(__doc__)
    if opts['--verbose']:
        level = logging.DEBUG
    else:
        level = logging.INFO
    logging.basicConfig(level=level,
                        format='%(asctime)s [%(levelname)s] %('
                               'filename)s:%(lineno)s %(funcName)s %('
                               'message)s',
                        datefmt="%Y-%m-%dT%H:%M:%S%z"
                        )
    try:
        line_number = 0
        for line_number, line in enumerate(get_log_entries(opts["LOG_FILE"])):
            pass
        else:
            print(f"Successfully processed {line_number} lines of log.")
    except Exception as e:
        logger.exception(f"Got error {e}")


if __name__ == "__main__":
    main()

src/metsuri/log_uploader.py
"""
Usage: log-uploader [options] LOG_FILE LOG_GROUP LOG_STREAM
       log-uploader --version


This will upload the given log to stream, each line as a log event. Every
line in the log should start with an '[<ISO8601-TIMESTAMP>]' to allow this
script to correctly add that data to logged data.

This script assumes we can write data faster towards AWS than it is being
generated by the serial logger to the log file(s).

Options:
  --watch              Stay on foreground and upload new events.
  --ignore-timestamp   Ignore stored timestamp and upload all events.
  --verbose            Enable verbose logging.
"""

import boto3
import botocore
import docopt
import logging
import os
import time
import datetime
import re
import itertools
from typing import Optional, NamedTuple, Iterable
import multiprocessing as mp
import queue
from pathlib import Path

TIMESTAMP_FILE_SUFFIX = ".lus"
AWS_MAX_BATCH_SIZE = 1048576
AWS_MAX_EVENT_TIME_SPAN = 24 * 3600

logger = logging.getLogger(__name__)

here = Path(__file__).parent.resolve()


class ChunkUploader:
    def __init__(self,
                 client,
                 group: str,
                 stream: str,
                 upload_sequence_token: str,
                 min_time_between_requests: float,
                 max_lines_in_batch: int,
                 max_time_between_uploads: float,
                 max_batch_size: int = AWS_MAX_BATCH_SIZE,
                 max_event_time_span: int = AWS_MAX_EVENT_TIME_SPAN,
                 timestamp_file_name: Optional[str] = None):
        self.batch = []
        self.min_time_between_requests = min_time_between_requests
        self.max_time_between_uploads = max_time_between_uploads
        self.max_event_time_span = max_event_time_span
        self.max_batch_size = max_batch_size
        self.prev_ts = time.time()
        self.client = client
        self.group = group
        self.stream = stream

        self.timestamp_file_name = timestamp_file_name
        self.upload_sequence_token = upload_sequence_token
        self._last_event_ts = None
        self._current_batch_size = 0
        self.max_lines_in_batch = max_lines_in_batch

    def append(self, event):
        # Timestamps need to be increasing in a batch.
        if self._last_event_ts and self._last_event_ts > event.timestamp:
            self.upload_current_batch()

        if self.batch and (event.timestamp - self.batch[0].timestamp) > datetime.timedelta(seconds=self.max_event_time_span):
            self.upload_current_batch()

        EVENT_HEADER_SIZE = 26
        event_size = len(event.message.encode('utf-8')) + EVENT_HEADER_SIZE
        if event_size + self._current_batch_size > self.max_batch_size:
            self.upload_current_batch()

        if event.message:
            self.batch.append(event)
            self._current_batch_size += event_size
        self._last_event_ts = event.timestamp

        self.maybe_upload()

    def maybe_upload(self):
        if len(self.batch) >= self.max_lines_in_batch:
            self.upload_current_batch()
        else:
            elapsed = self._elapsed()
            if elapsed and elapsed > self.max_time_between_uploads:
                self.upload_current_batch()

    def _update_timestamp(self, ts: datetime.datetime):
        if not self.timestamp_file_name:
            return

        try:
            with open(self.timestamp_file_name, "w") as fp:
                fp.write(ts.isoformat())
        except PermissionError:
            logger.debug("Can't write timestamp file")

    def upload_current_batch(self):
        if not self.batch:
            return

        self._maybe_throttle()
        logger.debug(f"Uploading batch")

        # From https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html
        #  The batch of events must satisfy the following constraints:
        #
        # The maximum batch size is 1,048,576 bytes. This size is
        # calculated as the sum of all event messages in UTF-8, plus 26
        # bytes for each log event.
        # None of the log events in the batch can be more than 2 hours in
        # the future.
        # None of the log events in the batch can be older than 14 days or
        # older than the retention period of the log group.
        # The log events in the batch must be in chronological order by
        # their timestamp. The timestamp is the time the event occurred,
        # expressed as the number of milliseconds after Jan 1,
        # 1970 00:00:00 UTC. (In AWS Tools for PowerShell and the AWS SDK
        # for .NET, the timestamp is specified in .NET format:
        # yyyy-mm-ddThh:mm:ss. For example, 2017-09-15T13:45:30.)
        # A batch of log events in a single request cannot span more than
        # 24 hours. Otherwise, the operation fails.
        # The maximum number of log events in a batch is 10,000.
        # There is a quota of 5 requests per second per log stream.
        # Additional requests are throttled. This quota can't be changed.

        self.upload_sequence_token = upload_batch(self.client, self.group,
                                                  self.stream, self.batch,
                                                  self.upload_sequence_token)
        self._update_timestamp(self._last_event_ts)
        self.batch = []
        self.prev_ts = time.time()
        self._current_batch_size = 0

    def _maybe_throttle(self):
        if self.prev_ts is not None:
            elapsed = self._elapsed()
            if elapsed < self.min_time_between_requests:
                throttle_seconds = self.min_time_between_requests - elapsed
                logger.info(
                    f"Throttling for {throttle_seconds}s to avoid hitting "
                    f"request quota")
                time.sleep(throttle_seconds)

    def _elapsed(self):
        if not self.prev_ts:
            return None
        else:
            elapsed = time.time() - self.prev_ts
            return elapsed


class EOF:
    pass


def log_entry_producer(q, log_filename, watch: bool = False, read_timestamp: bool = True):
    try:
        for ev in get_log_entries(log_filename, watch=watch, read_timestamp=read_timestamp):
            q.put(ev)
        q.put(EOF())
    except Exception as e:
        q.put(e)
        raise


def upload_log(name: str, group: str, stream: str, watch: bool = False,
               max_time_between_uploads: float = 5,
               min_time_between_requests: float = 0.2,
               max_lines_in_batch: int = 5000,
               max_batch_size: int = AWS_MAX_BATCH_SIZE,
               timestamp_file_name: Optional[str] = None,
               read_timestamp: bool = True):
    """

    :param name: Name of the log file.
    :param group: Log Group in AWS CloudWatch.
    :param stream: Log Stream in AWS CloudWatch.
    :param watch: Instead of returning when log file is at the end, keep reading and waiting for more input.
    :param max_time_between_uploads: If there is any input, avoid waiting longer than this to upload.
    :param min_time_between_requests: Wait at least this long between calls to avoid hitting request quota.
    :param max_lines_in_batch: Maximum number of events in a single batch.
    :param max_batch_size: Maximum size, in bytes, for a single batch.
    :param timestamp_file_name: Timestamp file name, leave as None for default.
    :param read_timestamp: If True, skip events while events have earlier time than the timestamp found in the `timestamp_file_name`.
    :return:
    """
    client = boto3.client('logs')

    upload_sequence_token = get_next_sequence_token(client, group, stream)

    if timestamp_file_name is None:
        timestamp_file_name = get_stamp_filename(name)
    uploader = ChunkUploader(client, group, stream, upload_sequence_token,
                             min_time_between_requests,
                             max_lines_in_batch,
                             max_time_between_uploads,
                             max_batch_size,
                             timestamp_file_name=timestamp_file_name)

    q = mp.Queue()
    producer = mp.Process(target=log_entry_producer, args=(q, name),
                          kwargs={'watch': watch, 'read_timestamp': read_timestamp})
    producer.daemon = True
    producer.start()

    while True:
        try:
            ev = q.get(block=True, timeout=max(max_time_between_uploads/5, 0.1))
            if isinstance(ev, EOF):
                logger.info("Got EOF")
                break
            elif isinstance(ev, Exception):
                logger.error(f"Reader encountered error; {ev}", exc_info=ev)
                break
            else:
                logger.debug(f"got line {ev.message}")
                uploader.append(ev)
        except queue.Empty:
            uploader.maybe_upload()
            if not producer.is_alive():
                logger.error("Producer seems to have died.")
                break

    logger.info("stopping")
    uploader.upload_current_batch()
    q.close()
    producer.join()


def upload_batch(client, group, stream, batch, token):
    rest = {}
    if token:
        rest['sequenceToken'] = token
    response = client.put_log_events(
        logGroupName=group,
        logStreamName=stream,
        logEvents=[dict(timestamp=int(ev.timestamp.timestamp() * 1000),
                        message=ev.message) for ev in batch],
        **rest)
    logger.info(f"Wrote {len(batch)} log events, rejections: "
                f"{response.get('rejectedLogEventsInfo', '<none>')}")
    return response.get('nextSequenceToken', "")


def get_next_sequence_token(client, group, stream):
    response = client.describe_log_streams(logGroupName=group,
                                           logStreamNamePrefix=stream)
    matching = list(filter(lambda xx: xx['logStreamName'] == stream, response['logStreams']))
    if matching:
        stream_description = matching[0]
        logger.debug(f"Stream description for {stream}: {stream_description}")
        sequence_token = stream_description.get('uploadSequenceToken', "")
    else:
        try:
            client.create_log_stream(logGroupName=group, logStreamName=stream)
        except botocore.exceptions.ClientError as e:
            logger.warning(f"Failed to create stream {stream} to log group {group}; {e}")
        sequence_token = ""
    return sequence_token


class Event(NamedTuple):
    timestamp: datetime.datetime
    message: str


def parse_log_line(line: Optional[str]) -> Optional[Event]:
    if line is None:
        return None

    m = re.search(r"(?P<timestamp>\d+-\d+-\d+[T ]\d{2}:\d{2}:\d{2}(\.\d+)?([+-]\d{2}:\d{2})?)]?:? ?(?P<msg>.*$)", line)
    if m is None:
        logger.warning(f"line \"{line}\" did not match when finding timestamp")
        return None

    timestamp = m.group("timestamp")
    try:
        dt = datetime.datetime.fromisoformat(timestamp)
    except ValueError:
        logger.warning(f"line \"{line}\" had an invalid timestamp.")
        return None

    if dt.tzinfo is None:
        # Treat as old format localtime log timestamp
        dt = dt.astimezone()
    msg = m.group("msg")
    return Event(timestamp=dt, message=msg)


def get_log_entries(name: str, watch: bool = False, read_timestamp: bool = True) -> Iterable[Event]:
    """
    Return a generator yielding log events. If `watch` is True, tracks file
    changes to continue reading from a newly created file with the same name.

    :param name:
    :param watch:
    :param read_timestamp:
    :return:
    """
    if read_timestamp:
        timestamp = get_timestamp(name)
    else:
        timestamp = None

    def get_lines(fp):
        line = "start"
        while line or watch:
            line = fp.readline()
            if line:
                yield line
            else:
                if watch:
                    # in case file is just being rotated.
                    if os.path.exists(name) and os.stat(name).st_ino != current_inode:
                        return
                    else:
                        time.sleep(0.5)
                        continue
                else:
                    return None

    while True:
        try:
            current_inode = os.stat(name).st_ino
            with open(name, errors="backslashreplace") as logfile:
                for event in itertools.dropwhile(
                        lambda arg: timestamp is not None and arg and arg.timestamp <= timestamp,
                        filter(None, (parse_log_line(line) for line in get_lines(logfile)))):
                    yield event
        except OSError as e:
            if e.errno == 2:
                time.sleep(0.5)
            else:
                raise

        if not watch:
            return


def get_timestamp(name) -> Optional[datetime.datetime]:
    state_file_name = get_stamp_filename(name)
    if os.path.exists(state_file_name):
        line = parse_log_line(open(state_file_name).read().strip())
        if line:
            return line.timestamp
        else:
            return None
    else:
        return None


def get_stamp_filename(name):
    return name + TIMESTAMP_FILE_SUFFIX


def output_version():
    print("Metsuri log collector, version: " + open(here / "VERSION").read().strip())


def main():
    mp.set_start_method('spawn')

    opts = docopt.docopt(__doc__)

    if opts['--version']:
        output_version()
        return

    if opts['--verbose']:
        level = logging.DEBUG
    else:
        level = logging.INFO
    logging.basicConfig(level=level,
                        format='%(asctime)s [%(levelname)s] %('
                               'filename)s:%(lineno)s %(funcName)s %('
                               'message)s',
                        datefmt="%Y-%m-%dT%H:%M:%S%z"
                        )

    upload_log(opts["LOG_FILE"], opts["LOG_GROUP"], opts["LOG_STREAM"],
               watch=opts['--watch'], read_timestamp=not opts['--ignore-timestamp'])
    logger.info("exiting")


if __name__ == "__main__":
    main()
